snap7/client_async.py

```diff
@@ -3,36 +3,131 @@
 """
 import asyncio
 import logging
-from ctypes import byref
+from ctypes import byref, Array
 
 from snap7.common import check_error
+from snap7.types import S7SZL
 from .client import Client
 
-logger = logging.getLogger(__name__)
+from concurrent.futures import ThreadPoolExecutor
+from functools import wraps, partial
+from typing import Union, Tuple, Awaitable
 
 
-def error_wrap(func):
-    """Parses a s7 error code returned the decorated function."""
-    async def f(*args, **kw):
-        code = await func(*args, **kw)
-        check_error(code, context="client")
+class ClientAsync(Client):
+    def __init__(self, max_workers=1):
+        super().__init__()
+        self.executor = ThreadPoolExecutor(max_workers=max_workers)
 
-    return f
+    def sync_to_async(self, func):
```

Comment:
That is a pretty smooth way to convert functions to coroutines tbh!
But in the end, aren't the Coroutines blocking other running Coroutines? Could be ProcessPoolExecutor be the better choice?

Reply:
As far as I know, I/O calls release GIL, so a thread should be enough. But maybe you're right. I'll look into it.

Reply:
I made a research and it seems that you are right about that. I/O bounded calls seem to be the exception in the whole Thread case,.